File: services/voice-agent-runtime/app/events.py
"""CloudEvents 1.0 envelope helpers (SPEC §4).

Envelope: {specversion, id, source, type, subject, time, tenantid (ext), data}.
Note: the booking-service command consumer requires `tenantid` to be the
tenant **UUID** and `subject` to be the tenant **slug**.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

SOURCE = "voice-agent-runtime"


def session_lifecycle_data(
    *,
    conversation_id: str,
    channel: str,
    site_slug: str,
    quality: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Data payload for SessionStarted/SessionEnded events.

    `quality` (SessionMetrics.quality_payload) is attached to SessionEnded
    only when the session produced data — the key stays absent otherwise so
    downstream consumers (crm-sync-service) can tell "no signals" apart from
    "zero values".
    """
    data: dict[str, Any] = {
        "conversationId": conversation_id,
        "channel": channel,
        "siteSlug": site_slug,
    }
    if quality:
        data["quality"] = quality
    return data


def new_cloudevent(
    *,
    type_: str,
    subject: str,
    tenant_uuid: str,
    data: dict[str, Any],
    event_id: str | None = None,
) -> dict[str, Any]:
    eid = event_id or str(uuid.uuid4())
    return {
        "specversion": "1.0",
        "id": eid,
        "source": SOURCE,
        "type": type_,
        "subject": subject,
        "time": datetime.now(timezone.utc)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z"),
        "tenantid": tenant_uuid,
        "data": data,
    }

File: services/payments-service/src/events.rs
//! CloudEvents 1.0 envelope (SPEC §4).

use chrono::Utc;
use serde::Serialize;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize)]
pub struct CloudEvent<T: Serialize> {
    pub specversion: &'static str,
    pub id: String,
    pub source: String,
    #[serde(rename = "type")]
    pub type_: String,
    pub subject: String,
    pub time: String,
    /// CloudEvents extension attribute carrying the tenant (SPEC §4).
    pub tenantid: String,
    pub data: T,
}

impl<T: Serialize> CloudEvent<T> {
    pub fn new(source: &str, type_: &str, subject: &str, tenant_id: &str, data: T) -> Self {
        Self {
            specversion: "1.0",
            id: Uuid::new_v4().to_string(),
            source: source.to_string(),
            type_: type_.to_string(),
            subject: subject.to_string(),
            time: Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
            tenantid: tenant_id.to_string(),
            data,
        }
    }
}
